File: matlab/table_1a.m
% Table B2 in Section B.4 in Canay, Illanes and Velez (2023)

% Inputs

% ../data/                               fake data
%  - A.csv               n  x (1+J0)     matrix of revenue differential
%  - D.csv               n  x (1+J)      matrix of all product portfolio
%  - J0.csv              J0 x 2          matrix of ownership by two firms

% G_restriction.m                        find test statistic and c. value
%  - m_function                          compute (26)-(27)
%  - m_hat                               compute a version of (38)
%  - cvalue_SN2S                         compute c. value as in (41)
%  - cvalue_EB2S                         compute c. value as in (48)

% output

% '_results/tables-tex/table_1a.tex'     confidence intervals and comp.time

% comment:
% the first column of A_matrix and D_matrix were used to index the markets,
% these first columns are useless in the rest of the code.

clc; clear all; close all;
addpath('1_functions');
mkdir('_results');

%% 1 Setup

% Import data
dgp = struct;
dgp.A_matrix = csvread(fullfile('..', 'data', 'A.csv'));
dgp.D_matrix = csvread(fullfile('..', 'data', 'D.csv'));
dgp.J0_vec = csvread(fullfile('..', 'data', 'J0.csv')); % (product, firm), where firm = 1 coca-cola, firm=2 energy-products

dgp.num_market = size(dgp.A_matrix, 1);
dgp.num_product = size(dgp.D_matrix, 1) - 1;
dgp.W_data = dgp.D_matrix(:, 2:end);

% Settings (cell arrays are used to loop over each of the four different specifications)
settings = struct;
settings.Vbar = {500, 500, 1000, 1000}; % Vbar is defined in Assumption 4.2 and appears in eq. (26)-(27).

settings.test_stat = {'CCK', 'CCK', 'CCK', 'CCK'}; % CCK as in eq. (38).
settings.cv = {'SN2S', 'EB2S', 'SN2S', 'EB2S'}; % Critical values as in eq. (41) and (47).
settings.alpha = {0.05, 0.05, 0.05, 0.05}; % significance level
settings.IV = {[], [], [], []}; % no IVs

% Technical settings (cell arrays are used to loop over the two parameters: theta1 and theta2)
sim = struct;
aux_grid1 = linspace(-40, 100, 141)';
aux_grid2 = linspace(-40, 100, 141)';
[grid1, grid2] = meshgrid(aux_grid1, aux_grid2);
sim.num_points = size(grid1, 1) * size(grid2, 2);
sim.grid_Theta = [reshape(grid1, [1 sim.num_points])', reshape(grid2, [1 sim.num_points])']; % to run in a personal laptop
sim.rng_seed = 20220826;
sim.num_boots = 1000;
sim.num_robots = 4; % number of parallel workers
sim.sim_name = 'table_1a';

results = struct;
results.CI_vec = {zeros(4, 2), zeros(4, 2)}; % specs x {LB, UB}
results.Tn_vec = zeros(size(sim.grid_Theta, 1), 4);
results.comp_time = zeros(4, 1);

%% 2 Computation
%  two steps:
%             i) compute test statistic and critical value
%            ii) construct confidence interevals

delete(gcp)
parpool('local', sim.num_robots)

for sim0 = 1:4

    disp(['case ' num2str(sim0)])

    tic

    reject_H = zeros(size(sim.grid_Theta, 1), 1); % save which points reject H0
    Test_vec = zeros(size(sim.grid_Theta, 1), 1);
    cv_vec = zeros(size(sim.grid_Theta, 1), 1);

    % Step 1: find test stat. Tn(theta) and c.value(theta) using G_restriction

    parfor point0 = 1:size(sim.grid_Theta, 1)
        theta0 = sim.grid_Theta(point0, :)';

        %test_H0: [T_n, c_value]
        test_H0 = G_restriction(dgp.W_data, dgp.A_matrix, theta0, dgp.J0_vec, settings.Vbar{sim0}, settings.IV{sim0}, 'all', settings.test_stat{sim0}, settings.cv{sim0}, settings.alpha{sim0}, sim.num_boots, sim.rng_seed);

        Test_vec(point0) = test_H0(1);
        cv_vec(point0) = test_H0(2);

        reject_H(point0) = 1 * (test_H0(1) > test_H0(2));
    end

    results.Tn_vec(:, sim0) = Test_vec;

    % Step 2: find confidence intervals using Tn(theta) and c.value(theta)

    % Confidence Interval for theta0 = (theta1,theta2)

    CS_vec = [];

    for point0 = 1:size(sim.grid_Theta, 1)
        theta0 = sim.grid_Theta(point0, :)';

        if reject_H(point0) == 0
            CS_vec = [CS_vec; theta0'];
        end

    end

    results.CS_unparti{sim0, 1} = CS_vec; % confidence interval without additional refinement on grid-search

    if sum(size(CS_vec)) == 0 % it may be the CI is empty
        results.CI_vec{1}(sim0, :) = [NaN NaN];
        results.CI_vec{2}(sim0, :) = [NaN NaN];
        [~, point0] = min(Test_vec);
        theta0 = sim.grid_Theta(point0, :);
        results.CI_vec{1}(sim0, 2) = theta0(1); % in this case, we report [nan, argmin test statistic]
        results.CI_vec{2}(sim0, 2) = theta0(2); % in this case, we report [nan, argmin test statistic]
    end

    % Confidence Interval for theta0 after grid-search refinement

    if sum(size(CS_vec)) > 0

        aux1 = linspace(min(CS_vec(:, 1)) + 1, max(CS_vec(:, 1)) - 1, max(CS_vec(:, 1)) - min(CS_vec(:, 1)) - 1);
        aux1 = [aux1 linspace(max(CS_vec(:, 1)) - 0.9, max(CS_vec(:, 1)) + 0.9, 19)];
        aux1 = [linspace(min(CS_vec(:, 1)) - 0.9, min(CS_vec(:, 1)) + 0.9, 19) aux1];

        aux2 = linspace(min(CS_vec(:, 2)) + 1, max(CS_vec(:, 2)) - 1, max(CS_vec(:, 2)) - min(CS_vec(:, 2)) - 1);
        aux2 = [aux2 linspace(max(CS_vec(:, 2)) - 0.9, max(CS_vec(:, 2)) + 0.9, 19)];
        aux2 = [linspace(min(CS_vec(:, 2)) - 0.9, min(CS_vec(:, 2)) + 0.9, 19) aux2];

        [grid1, grid2] = meshgrid(aux1, aux2);
        num_pointsv2 = size(grid1, 1) * size(grid2, 2);
        grid_searchv2 = [reshape(grid1, [1 num_pointsv2])', reshape(grid2, [1 num_pointsv2])']; % to run in a personal laptop

        reject_H2 = zeros(num_pointsv2, 1); % to save which points reject H0

        parfor point0 = 1:num_pointsv2
            theta0 = grid_searchv2(point0, :)';

            %test_H0: [T_n, c_value]
            test_H0 = G_restriction(dgp.W_data, dgp.A_matrix, theta0, dgp.J0_vec, settings.Vbar{sim0}, settings.IV{sim0}, 'all', settings.test_stat{sim0}, settings.cv{sim0}, settings.alpha{sim0}, sim.num_boots, sim.rng_seed);

            reject_H2(point0) = 1 * (test_H0(1) > test_H0(2));
        end

        CS_vec = [];

        for point0 = 1:num_pointsv2
            theta0 = grid_searchv2(point0, :)';

            if reject_H2(point0) == 0
                CS_vec = [CS_vec; theta0'];
            end

        end

        results.CS_unparti{sim0, 2} = CS_vec;

        results.CI_vec{1}(sim0, :) = [max(min(CS_vec(:, 1)), -40) min(max(CS_vec(:, 1)), 100)];
        results.CI_vec{2}(sim0, :) = [max(min(CS_vec(:, 2)), -40) min(max(CS_vec(:, 2)), 100)];

    end

    toc
    time = toc;
    results.comp_time(sim0, 1) = time;
end

%% 3 Save results
save(fullfile('_results', strcat(sim.sim_name, '.mat')), 'dgp', 'settings', 'sim', 'results');

%% 4 Print table
cd_name = 'tables-tex';
mkdir(fullfile('_results', cd_name));

f = fopen(fullfile('_results', cd_name, strcat(sim.sim_name, '.tex')), 'w'); % Open file for writing

fprintf(f, '%s\n', '\begin{tabular}{c c c c c}');
fprintf(f, '%s\n', '\hline \hline');
fprintf(f, '%s\n', '~ & Crit. Value & $\theta_1$: Coca-Cola &$\theta_2$: Energy Brands & Comp. Time \\');
fprintf(f, '%s\n', '\hline');

for row0 = 1:4

    if strcmp(settings.cv{row0}, 'SN2S')
        Vbar = ['$\Bar{V}$=' num2str(settings.Vbar{row0})];
        cvalue0 = 'self-norm';
    elseif strcmp(settings.cv{row0}, 'EB2S')
        Vbar = "~";
        cvalue0 = 'bootstrap';
    end

    fprintf(f, '%s%s%s%s%5.1f%s%5.1f%s%5.1f%s%5.1f%s%5.1f%s\n', ...
        Vbar, ' & ', cvalue0, ' & [', ...
        results.CI_vec{1}(row0, 1), ' , ', results.CI_vec{1}(row0, 2), '] & [', ...
        results.CI_vec{2}(row0, 1), ' , ', results.CI_vec{2}(row0, 2), '] &', ...
        results.comp_time(row0, 1), '\\');

    if row0 == 2
        fprintf(f, '%s\n', '\hline');
    end

end

fprintf(f, '%s\n', '\hline \hline');
fprintf(f, '%s', '\end{tabular}');

File: r/table_1a.R
# Table 1 in Section 8.1 in Canay, Illanes and Velez (2023)

# Import packages and functions
require(readr)
require(tictoc)
require(Rfast)
require(Rfast2)
require(foreach)
require(doParallel)
require(xtable)
require(devtools)

# Install and import package
devtools::install("ineqfunctions", upgrade = "never")

if (!dir.exists("_results")) {
  dir.create("_results")
}

# Import data
datasets <- c("A", "D", "J0")
data_path <- file.path("..", "data")
dgp <- sapply(datasets, function(dataset) {
  unname(as.matrix(readr::read_csv(
    file.path(data_path, paste0(dataset, ".csv")),
    col_names = F,
    show_col_types = F
  )))
}, simplify = F)
dgp$W <- dgp$D[, -1]
# number of markets
n <- nrow(dgp$A)

# Settings (cell arrays are used to loop over each of the four different specifications)
settings <- list(
  # v_bar is defined in Assumption 4.2 and appears in eq. (26)-(27).
  v_bar = c(500, 500, 1000, 1000),
  # CCK as in eq. (38).
  test_stat = rep("CCK", 4),
  # Critical values as in eq. (41) and (47).
  cv = rep(c("SN2S", "EB2S"), 2),
  # significance level
  alpha = 0.05,
  # no IVs
  iv = NULL
)

# Technical settings (lists are used to loop over the two parameters: theta1 and theta2)
sim <- list(
  grid_theta = expand.grid(
    seq.int(-40, 100),
    seq.int(-40, 100)
  ),
  rng_seed = 20220826,
  bootstrap_replications = 1000,
  num_robots = 4,
  sim_name = "table_1a"
)
# number of theta values: 19881
sim$grid_size <- nrow(sim$grid_theta)
# dimension of theta: 2
sim$dim_theta <- ncol(sim$grid_theta)

results <- list(
  ci_vector = list(matrix(NA, 4, 2), matrix(NA, 4, 2)),
  tn_vector = lapply(sim$grid_theta, function(grid) {
    matrix(NA, nrow = length(grid), ncol = 4)
  }),
  comp_time = rep(NA, 4)
)

# Generate bootstrap indices
# number of markets
bootstrap_indices <- get_bootstrap_indices(
  n, sim$bootstrap_replications, sim$rng_seed
)

# Parallel computing
cl <- parallel::makePSOCKcluster(sim$num_robots)
doParallel::registerDoParallel(cl)

## 2 Computation
#  two steps:
#             i) compute test statistic and critical value
#            ii) conlist() confidence interevals

for (sim_i in 1:4) {
  tictoc::tic(paste("Simulation", sim_i))

  # Step 1: find test stat. Tn(theta) and c.value(theta) using g_restriction
  output <- foreach::foreach(
    i = 1:sim$grid_size,
    .combine = "rbind"
  ) %dopar% {
    theta <- as.numeric(sim$grid_theta[i, ])

    # output: [T_n, c_value]
    ineqfunctions::g_restriction(
      theta = theta,
      w_data = dgp$W,
      a_matrix = dgp$A,
      j0_vector = dgp$J0,
      v_bar = settings$v_bar[sim_i],
      alpha = settings$alpha,
      grid0 = "all",
      test0 = settings$test_stat[sim_i],
      cvalue = settings$cv[sim_i],
      iv_matrix = settings$iv,
      bootstrap_indices = bootstrap_indices,
    )
  }

  test_vec <- output[, 1]
  cv_vec <- output[, 2]

  # Theta values for which the null is not rejected
  cs_vec <- sim$grid_theta[test_vec <= cv_vec, ]


  for (theta_index in 1:sim$dim_theta) {
    # Create results objects
    results$tn_vector[[theta_index]][, sim_i] <- test_vec

    if (nrow(cs_vec) == 0) {
      # it may be the CI is empty
      # in this case, we report [nan, argmin test statistic]
      results$ci_vector[[theta_index]][sim_i, ] <- c(
        NaN,
        sim$grid_theta[which.min(test_vec), theta_index]
      )
    } else {
      results$ci_vector[[theta_index]][sim_i, ] <- c(
        min(cs_vec[, theta_index]),
        max(cs_vec[, theta_index])
      )
    }
  }

  # Stop the timer
  temp_timer <- tictoc::toc()
  results$comp_time[sim_i] <- temp_timer$toc - temp_timer$tic
}

# Stop parallel computing
stopImplicitCluster()

## 3 Save results
save(results, file = file.path("_results", paste0(sim$sim_name, ".Rdata")))


# save(fullfile('_results', strcat(sim[['sim_name']], '.mat')), 'dgp', 'settings', 'sim', 'results')
#
## 4 Print table
table_dir <- file.path("_results", "tables-tex")
if (!dir.exists(table_dir)) {
  dir.create(table_dir)
}

# Format CI as [lb, ub]
formatted_ci <- sapply(results$ci_vector, function(ci_theta) {
  apply(ci_theta, 1, function(x) {
    paste0("[", sprintf("%.1f", x[1]), ", ", sprintf("%.1f", x[2]), "]")
  })
})

# Make table as matrix
the_table <- cbind(
  settings$v_bar,
  settings$cv,
  formatted_ci,
  sprintf("%.2f", results$comp_time)
)

# Add colnames
colnames(the_table) <- c(
  "$\\Bar{V}$",
  "Crit. Value",
  "$\\theta_1$: Coca-Cola",
  "$\\theta_2$: Energy Brands",
  "Comp. Time"
)

# Save the table
print(
  xtable(the_table, digits = 4),
  include.rownames = FALSE,
  sanitize.colnames.function = identity,
  file = file.path(table_dir, paste0(sim$sim_name, ".tex"))
)
